packages/network-enablement-controller/src/NetworkEnablementController.ts

```diff
@@ -0,0 +1,353 @@
+
+export type NetworkEnablementControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  NetworkEnablementControllerActions | AllowedActions,
+  NetworkEnablementControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * Gets the default state for the NetworkEnablementController.
+ *
+ * @returns The default state with pre-enabled networks.
+ */
+const defaultState = (): NetworkEnablementControllerState => ({
+  enabledNetworkMap: {
+    [KnownCaipNamespace.Eip155]: {
+      [ChainId[BuiltInNetworkName.Mainnet]]: true,
+      [ChainId[BuiltInNetworkName.LineaMainnet]]: true,
+      [ChainId[BuiltInNetworkName.BaseMainnet]]: true,
+    },
+    [KnownCaipNamespace.Solana]: {
+      [SolScope.Mainnet]: false,
+    },
+  } as EnabledMap,
+});
+
+// Metadata for the controller state
+const metadata = {
+  enabledNetworkMap: {
+    persist: true,
+    anonymous: true,
+  },
+};
+
+const mergeEnabledMaps = (
+  base: EnabledMap,
+  override?: Partial<EnabledMap>,
+): EnabledMap => {
+  if (!override) {
+    return base;
+  }
+  return (Object.keys(base) as CaipNamespace[]).reduce((acc, ns) => {
+    acc[ns] = { ...base[ns], ...(override[ns] ?? {}) };
+    return acc;
+  }, {} as EnabledMap);
+};
+
+export class NetworkEnablementController extends BaseController<
+  typeof controllerName,
+  NetworkEnablementControllerState,
+  NetworkEnablementControllerMessenger
+> {
+  /**
+   * Creates a NetworkEnablementController instance.
+   *
+   * @param args - The arguments to this function.
+   * @param args.messenger - Messenger used to communicate with BaseV2 controller.
+   * @param args.state - Initial state to set on this controller.
+   */
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: NetworkEnablementControllerMessenger;
+    state?: Partial<NetworkEnablementControllerState>;
+  }) {
+    // Call the constructor of BaseControllerV2
+    super({
+      messenger,
+      metadata,
+      name: controllerName,
+      state: {
+        enabledNetworkMap: mergeEnabledMaps(
+          defaultState().enabledNetworkMap,
+          state?.enabledNetworkMap,
+        ),
+      },
+    });
+
+    this.messagingSystem = messenger;
+
+    messenger.subscribe('NetworkController:networkAdded', ({ chainId }) => {
+      this.#ensureNetworkEntry(chainId, false);
+      this.#toggleNetwork(chainId, true);
+    });
+
+    messenger.subscribe('NetworkController:networkRemoved', ({ chainId }) => {
+      this.#removeNetworkEntry(chainId);
+    });
+  }
+
+  setEnabledNetwork(chainId: Hex | CaipChainId): void {
+    this.#toggleNetwork(chainId, true);
+  }
+
+  setDisabledNetwork(chainId: Hex | CaipChainId): void {
+    this.#toggleNetwork(chainId, false);
+  }
+
+  isNetworkEnabled(chainId: Hex | CaipChainId): boolean {
+    try {
+      const { namespace, storageKey } = this.#deriveKeys(chainId);
+      return Boolean(this.state.enabledNetworkMap[namespace]?.[storageKey]);
```

Comment:
Nit: Instead of using `Boolean` (which has unexpected behavior) what do you think about being explicit here? 

``` suggestion
      return (
        namespace in this.state.enabledNetworkMap &&
        this.state.enabledNetworkMap[namespace][storageKey]
      );